flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/hashtable/BaseHybridHashTable.java

```diff
@@ -516,4 +511,10 @@ public static int hash(int hashCode, int level) {
 		return code >= 0 ? code : -(code + 1);
 	}
 
+	/**
+	 * Partition level hash again, for avoid two layer hash conflict.
+	 */
+	static int partitionLevelHash(int hash) {
+		return hash ^ (hash >>> 16);
+	}
```

Comment:
@JingsongLi Thanks a lot for bringing this up. 
I agree with you that this is computationally efficient. 

However, according to my experince, this may not be a good hash function. In practice, we usually encounter small integers (with most high bits equal to 0), so we have hash >>> 16 == 0, and hash ^ (hash >>> 16) will be equal to hash.

IMO, hash ^ (hash << 16) is much better, as it maps small integers uniformly into the space of all integers. 

That being that, this is just my personal suggestion. The final choice should be dependent on the real scenario.

Reply:
Thanks for involving this.
> with most high bits equal to 0

Good point, but it should not. 
- In `BinaryHashTable`: HashCode is from murmurhash, it is good.
- In `LongHybridHashTable`: HashCode is from `hashLong`, it is good too.

`hash ^ (hash >>> 16)` is just from `HashMap.hash`, I think it is OK.